DataFormats/Detectors/FIT/FT0/include/DataFormatsFT0/RawEventData.h

```diff
@@ -0,0 +1,149 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+//
+//file RawEventData.h class  for RAW data format
+//Alla.Maevskaya@cern.ch
+// with Artur.Furs
+//
+#ifndef ALICEO2_FIT_RAWEVENTDATA_H_
+#define ALICEO2_FIT_RAWEVENTDATA_H_
+
+#include "DataFormatsFT0/Digit.h"
+#include "Headers/RAWDataHeader.h"
+#include "DataFormatsFT0/LookUpTable.h"
+#include <CommonDataFormat/InteractionRecord.h>
+#include <iostream>
+#include <bitset>
+#include "Rtypes.h"
+namespace o2
+{
+namespace ft0
+{
+constexpr int NCHANNELS_FT0 = 208;
+constexpr int NCHANNELS_PM = 12;
+constexpr int NPMs = 18;
+constexpr int NBITS_EVENTDATA = 9;
+
+struct EventHeader {
+  ushort startDescriptor : 4;
+  uint reservedField : 28;
+  uint orbit : 32;
+  short bc : 12;
+  short nGBTWords : 4;
+  ClassDefNV(EventHeader, 1);
+};
+struct EventData {
+  short int time : 12;
+  short int charge : 12;
+  unsigned short int numberADC : 1;
+  bool isDoubleEvent : 1;
+  bool is1TimeLostEvent : 1;
+  bool is2TimeLostEvent : 1;
+  bool isADCinGate : 1;
+  bool isTimeInfoLate : 1;
+  bool isAmpHigh : 1;
+  bool isEventInTVDC : 1;
+  bool isTimeInfoLost : 1;
+  uint channelID : 4;
+  ClassDefNV(EventData, 1);
+};
+
+class RawEventData
+{
+ public:
+  RawEventData() = default;
+  //virtual ~RawEventData();
+  void GenerateHeader(int nChannels);
+  void GenerateData();
+  void GenerateRandomHeader(int nChannels);
+  void GenerateRandomData();
+  void GenerateRandomEvent(int nChannels);
+  EventHeader* GetEventHeaderPtr() { return &mEventHeader; }
+  EventData* GetEventDataPtr() { return mEventData; }
+  void Print(bool doPrintData = false);
+  enum EEventDataBit { kNumberADC,
+                       kIsDoubleEvent,
+                       kIs1TimeLostEvent,
+                       kIs2TimeLostEvent,
+                       kIsADCinGate,
+                       kIsTimeInfoLate,
+                       kIsAmpHigh,
+                       kIsEventInTVDC,
+                       kIsTimeInfoLost };
+  const static int gStartDescriptor = 0x0000000f;
+
+  int size() const
+  {
+    return 4                         // RAWDataHeader
+           + 1                       // EventHeader
+           + mEventHeader.nGBTWords; // EventData
+  }
+
+  void write(std::ostream& str)
+  {
+    str.write(reinterpret_cast<const char*>(&mRDH), sizeof(mRDH));
+    str.write(reinterpret_cast<const char*>(&mEventHeader), sizeof(mEventHeader));
+    char padding[16 - sizeof(mEventHeader)] = {};
+    if (mIsPadded)
+      str.write(padding, sizeof(padding));
+    //  static_assert(sizeof(mEventHeader) == 2 * sizeof(mEventData[0]), "This code assumes that pairs of data and header require the same padding");
+    for (int i = 0; i < mEventHeader.nGBTWords; ++i) {
+      str.write(reinterpret_cast<const char*>(&mEventData[2 * i]), sizeof(mEventData[0]) * 2);
+      if (mIsPadded)
+        str.write(padding, sizeof(padding));
+    }
```

Comment:
Using sizeof here is wrong (see comment on EventHeader and EventData). 
Also, ``str.write(reinterpret_cast<const char*>(&mEventData[2 * i]), sizeof(mEventData[0]) * 2);`` will write 16 bytes, overflowing GBT word size.
If you modify these structs as I suggested, then you can write as:
````
void write(std::ostream& str)
{
  constexpr int CRUWordSize = 16;
  const char padding[CRUWordSize] = {0};

  str.write(reinterpret_cast<const char*>(&mRDH), sizeof(mRDH));
  str.write(reinterpret_cast<const char*>(&mEventHeader), EventHeader::PayloadSize );
  if (mIsPadded) {
    str.write(padding, CRUWordSize - EventHeader::PayloadSize);
  }
  for (int i = 0; i < mEventHeader.nGBTWords; ++i) {
    str.write(reinterpret_cast<const char*>(&mEventData[2*i]), EventData::PayloadSize);
    str.write(reinterpret_cast<const char*>(&mEventData[2*i+1]), EventData::PayloadSize);
    if (mIsPadded) {
      str.write(padding, CRUWordSize - 2*EventData::PayloadSize);
    }
  }
}
````

````

````